Quarkus master can't handle space in path for native mode, 1.4.1.Final can

Quarkus master can't handle space in path for native mode, 1.4.1.Final can handle space in path

Build generates native app but it is not working and immediatelly ends with exit code 1.

Reproducer:
```bash
rm -rf directory\ with\ spaces
mkdir directory\ with\ spaces
cd directory\ with\ spaces
mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started \
    -DplatformArtifactId=quarkus-bom \
    -DclassName="org.acme.getting.started.GreetingResource" \
    -Dpath="/hello"
cd getting-started

mvn clean verify -Dnative  #FAIL
mvn clean verify -Dnative -Dquarkus.platform.version=1.4.1.Final -Dquarkus-plugin.version=1.4.1.Final # OK
```

```
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] /Users/rsvoboda/.sdkman/candidates/java/20.0.0.r11-grl/bin/native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=1 -J-Duser.language=en -J-Dfile.encoding=UTF-8 --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime -H:+JNI -jar getting-started-1.0-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:-IncludeAllTimeZones -H:EnableURLProtocols=http --no-server -H:-UseServiceLoaderFeature -H:+StackTrace getting-started-1.0-SNAPSHOT-runner
[getting-started-1.0-SNAPSHOT-runner:84103]    classlist:   4,022.02 ms,  1.44 GB
[getting-started-1.0-SNAPSHOT-runner:84103]        (cap):   2,818.60 ms,  1.44 GB
[getting-started-1.0-SNAPSHOT-runner:84103]        setup:   4,304.40 ms,  1.44 GB
java.lang.RuntimeException: Failed to load resource META-INF/resources/index.html
	at io.quarkus.runtime.ResourceHelper.registerResources(ResourceHelper.java:25)
	at io.quarkus.runner.AutoFeature.beforeAnalysis(AutoFeature.zig:791)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$7(NativeImageGenerator.java:674)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:63)
	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:674)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:530)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:445)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.IllegalArgumentException: Failed to translate file:/Users/rsvoboda/Downloads/directory with spaces/getting-started/target/getting-started-1.0-SNAPSHOT-native-image-source-jar/getting-started-1.0-SNAPSHOT-runner.jar to local path
	at io.quarkus.runtime.util.ClassPathUtils.toLocalPath(ClassPathUtils.java:215)
	at io.quarkus.runtime.util.ClassPathUtils.readStream(ClassPathUtils.java:186)
	at io.quarkus.runtime.util.ClassPathUtils.consumeStream(ClassPathUtils.java:159)
	at io.quarkus.runtime.util.ClassPathUtils.consumeAsStreams(ClassPathUtils.java:53)
	at io.quarkus.runtime.ResourceHelper.registerResources(ResourceHelper.java:17)
	... 12 more
Caused by: java.net.URISyntaxException: Illegal character in path at index 40: file:/Users/rsvoboda/Downloads/directory with spaces/getting-started/target/getting-started-1.0-SNAPSHOT-native-image-source-jar/getting-started-1.0-SNAPSHOT-runner.jar
	at java.base/java.net.URI$Parser.fail(URI.java:2913)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3084)
	at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3166)
	at java.base/java.net.URI$Parser.parse(URI.java:3114)
	at java.base/java.net.URI.<init>(URI.java:600)
	at java.base/java.net.URL.toURI(URL.java:1035)
	at io.quarkus.runtime.util.ClassPathUtils.toLocalPath(ClassPathUtils.java:213)
	... 16 more
[getting-started-1.0-SNAPSHOT-runner:84103]   (typeflow):   9,091.29 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]    (objects):   7,238.78 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]   (features):     372.02 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]     analysis:  17,238.98 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]     (clinit):     356.23 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]     universe:     991.22 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]      (parse):   1,545.88 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]     (inline):   2,383.63 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]    (compile):  12,152.28 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]      compile:  17,064.11 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]        image:   2,123.82 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]        write:     544.87 ms,  2.90 GB
[getting-started-1.0-SNAPSHOT-runner:84103]      [total]:  46,609.87 ms,  2.90 GB
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 49210ms
```

--- a/.github/workflows/ci-actions.yml
+++ b/.github/workflows/ci-actions.yml
@@ -422,7 +422,7 @@ jobs:
               virtual-http
               rest-client
           - category: Misc1
-            timeout: 40
+            timeout: 45
             test-modules: >
               maven
               jackson
@@ -526,11 +526,18 @@ jobs:
       - name: Build with Maven
         env: 
           TEST_MODULES: ${{matrix.test-modules}}
+          CATEGORY: ${{matrix.category}}
         run: |
           for i in $TEST_MODULES
           do modules+=("integration-tests/$i"); done
           IFS=,
           eval mvn -pl "${modules[*]}" $NATIVE_TEST_MAVEN_OPTS
+          # add the 'simple with spaces' project to the run of 'Misc1' by executing it explicitly
+          # done because there is no good way to pass strings with empty values to the previous command
+          # so this hack is as good as any
+          if [ "$CATEGORY" == "Misc1" ]; then
+            mvn -Dnative -Dquarkus.native.container-build=true -B --settings .github/mvn-settings.xml -f 'integration-tests/simple with space/' verify
+          fi
       - name: Prepare failure archive (if maven failed)
         if: failure()
         shell: bash

--- a/core/runtime/src/main/java/io/quarkus/runtime/util/ClassPathUtils.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/util/ClassPathUtils.java
@@ -175,17 +175,21 @@ public class ClassPathUtils {
      */
     public static <R> R readStream(URL url, Function<InputStream, R> function) throws IOException {
         if (JAR.equals(url.getProtocol())) {
-            final URI uri;
-            try {
-                uri = new URI(url.toString());
-            } catch (URISyntaxException e) {
-                throw new IOException(e);
-            }
+            final URI uri = toURI(url);
             final String file = uri.getSchemeSpecificPart();
-            final int exclam = file.lastIndexOf('!');
-            final Path jar = toLocalPath(exclam >= 0 ? new URL(file.substring(0, exclam)) : url);
+            final int fileExclam = file.lastIndexOf('!');
+            final URL jarURL;
+            if (fileExclam > 0) {
+                // we need to use the original url instead of the scheme specific part because it contains the properly encoded path
+                String urlFile = url.getFile();
+                int urlExclam = urlFile.lastIndexOf('!');
+                jarURL = new URL(urlFile.substring(0, urlExclam));
+            } else {
+                jarURL = url;
+            }
+            final Path jar = toLocalPath(jarURL);
             try (FileSystem jarFs = FileSystems.newFileSystem(jar, (ClassLoader) null)) {
-                try (InputStream is = Files.newInputStream(jarFs.getPath(file.substring(exclam + 1)))) {
+                try (InputStream is = Files.newInputStream(jarFs.getPath(file.substring(fileExclam + 1)))) {
                     return function.apply(is);
                 }
             }
@@ -198,6 +202,16 @@ public class ClassPathUtils {
         try (InputStream is = url.openStream()) {
             return function.apply(is);
         }
+    }
+
+    private static URI toURI(URL url) throws IOException {
+        final URI uri;
+        try {
+            uri = new URI(url.toString());
+        } catch (URISyntaxException e) {
+            throw new IOException(e);
+        }
+        return uri;
     }
 
     /**

--- a/integration-tests/pom.xml
+++ b/integration-tests/pom.xml
@@ -106,6 +106,7 @@
         <module>reactive-messaging-amqp</module>
         <module>rest-client</module>
         <module>packaging</module>
+        <module>simple with space</module>
     </modules>
 
     <build>

--- a/integration-tests/simple with space/pom.xml
+++ b/integration-tests/simple with space/pom.xml
@@ -1,0 +1,92 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>quarkus-integration-tests-parent</artifactId>
+        <groupId>io.quarkus</groupId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>quarkus-integration-test-simple-with-space</artifactId>
+    <name>Quarkus - Integration Tests - Simple with space</name>
+    <description>
+        A very basic integration test that contains a space in the directory and also resources files that contain spaces
+    </description>
+    <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy</artifactId>
+        </dependency>
+        <!-- added because /META-INF/resources/licenses/ contains files with spaces in their names -->
+        <dependency>
+            <groupId>org.primefaces</groupId>
+            <artifactId>primefaces</artifactId>
+            <version>8.0</version>
+        </dependency>
+
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>rest-assured</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>build</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
+    <profiles>
+        <profile>
+            <id>native</id>
+            <activation>
+                <property>
+                    <name>native</name>
+                </property>
+            </activation>
+            <build>
+                <plugins>
+                    <plugin>
+                        <artifactId>maven-failsafe-plugin</artifactId>
+                        <executions>
+                            <execution>
+                                <goals>
+                                    <goal>integration-test</goal>
+                                    <goal>verify</goal>
+                                </goals>
+                                <configuration>
+                                    <systemProperties>
+                                        <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
+                                    </systemProperties>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+            <properties>
+                <quarkus.package.type>native</quarkus.package.type>
+            </properties>
+        </profile>
+    </profiles>
+
+</project>

--- a/integration-tests/simple with space/src/main/java/io/quarkus/it/spaces/GreetingResource.java
+++ b/integration-tests/simple with space/src/main/java/io/quarkus/it/spaces/GreetingResource.java
@@ -1,0 +1,21 @@
+package io.quarkus.it.spaces;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import org.primefaces.util.Constants;
+
+@Path("/hello")
+public class GreetingResource {
+
+    // make sure we reference something from Primefaces so GraalVM doesn't throw out the entire jar
+    private static final String PRIMEFACES_DOWNLOAD_COOKIE = Constants.DOWNLOAD_COOKIE;
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return "hello";
+    }
+}

--- a/integration-tests/simple with space/src/test/java/io/quarkus/it/spaces/GreetingResourceTest.java
+++ b/integration-tests/simple with space/src/test/java/io/quarkus/it/spaces/GreetingResourceTest.java
@@ -1,0 +1,22 @@
+package io.quarkus.it.spaces;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.CoreMatchers.is;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class GreetingResourceTest {
+
+    @Test
+    public void testHelloEndpoint() {
+        given()
+                .when().get("/hello")
+                .then()
+                .statusCode(200)
+                .body(is("hello"));
+    }
+
+}

--- a/integration-tests/simple with space/src/test/java/io/quarkus/it/spaces/NativeGreetingResourceIT.java
+++ b/integration-tests/simple with space/src/test/java/io/quarkus/it/spaces/NativeGreetingResourceIT.java
@@ -1,0 +1,9 @@
+package io.quarkus.it.spaces;
+
+import io.quarkus.test.junit.NativeImageTest;
+
+@NativeImageTest
+public class NativeGreetingResourceIT extends GreetingResourceTest {
+
+    // Execute the same tests but in native mode.
+}

